### DISCUSSION TITLE
Is using introspect in the new migration workflow considered ok?

### DISCUSSION THREAD

**Discussion by kswope** [prisma/prisma#4603] [Q&A] [answered]
Honestly, I don't want to ever modify the schema file by hand and deal with *yet another syntax in my life*.  Thing is, prisma lends itself well to using gui apps like DBeaver and Postico.

What seems to work is

1. modify database with Postico
2. npx prisma introspect
3. npx prisma migrate dev --preview-feature
4. "Drift detected: Your database schema is not in sync with your migration history." Do you want to continue? … yes
5. Name of migration ›
6. Everything is now in sync.

I don't see this flow in the docs but I'm thinking it would be very popular.  

Will this go horribly wrong eventually?

Is introspect output relatively deterministic, like table/field orders, capitalizations, etc, so it doesn't make a mess with versioning?

**Comment by ryands17** [comment]
Hey @kswope 👋 

Are the instructions you specified correct? The above doesn't work for me.

After running introspection and then `prisma migrate dev --preview-feature`, I get the following error:
![image](https://user-images.githubusercontent.com/19697099/101876184-05a74d80-3bb2-11eb-98d6-63a75c5c863f.png)

**Reply by kswope** [reply]
Yes I copied and pasted everything.  This is just a newly created tutorial project.  Maybe there's some difference between that and what you are working with.  This is what I set it up with https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch-typescript-postgres

I've been using prisma for about two days, so I'm just doing the tutorials right now

**Comment by tomhoule** [comment, marked as answer]
Hi @kswope — this is not the workflow that migrate preview is designed for, but it should mostly work. You will however lose information with the introspection steps: anything that isn't represented in the prisma schema (check constraints, views, etc.), and importantly, constraint names, which means some of the migrations we generate (like drop foreign keys) wouldn't work if the constraint names don't match between your reference database and your migration files.

This is a workflow we could do more to support, though, I think it's worth exploring further. If you have the time, can you create an issue to discuss this use-case and workflows built around it?

**Reply by carlos-moreira** [reply]
@tomhoule I have a question about this matter: The correct workflow of migrate preview is editing the schema.prisma file manually, and after that running:
```prisma migrate dev --preview-feature```
to generate the new migration file and get your database up to date with the migration, right?

I think it would be a lot easier to be able to, instead of editing the schema.prisma file manually, just update your database schema using SQL, and after that re-running prisma introspect to automatically update the schema.prisma file with only the changes that were added to the database, what is much less error prone than editing the file manually.

After that I would just run ```prisma migrate dev --preview-feature``` again to get the new migration file. The problem here is that prisma migrate would try to run the migration file and it would give an error message telling the changes were already made on the database because I had already applied the changes before running prisma introspect. For this new workflow to work we would need a way to tell prisma migrate that the changes were already made, but the table _prisma_migrations would need to be filled with this information, maybe doing the steps described here: [https://www.prisma.io/docs/guides/prisma-guides/add-prisma-migrate-to-a-project#baseline-your-production-environment](https://www.prisma.io/docs/guides/prisma-guides/add-prisma-migrate-to-a-project#baseline-your-production-environment)

---EDIT

This workflow just worked for me:

1. Edit database with SQL.
2. Run ```prisma introspect``` to update the _schema.prisma_ file automatically, keeping the git history diff only with the changes that were really made.
3. Run ```prisma migrate dev --create-only --preview-feature``` to create the migration but not actually running it.
4. Mark the migration as applied using the command ```prisma migrate resolve --applied "migration-id" --preview-feature```, this is possible because you already "applied" this migration on step 1.
5. Commit the changes to your git repo and your team can apply this new migration normally.

I think this work pretty well for my use case, and I think this could be added as a new workflow on the docs, is there anything wrong with these steps? @tomhoule

**Reply by albertoperdomo** [reply]
I would be wary of committing migrations that you have not applied yourself. You could be committing failing migrations (1) or migrations that do not lead the same result potentially and this could cause issues where you and the rest of the team have schemas that are not 100% in sync.

Generating migrations from changes in PSL has some limitations. An example of 1 is:

you adding a new required field without specifying a default in the Prisma schema. This cannot be dealt with automatically without user intervention. At this point you can either a) add a default to the Prisma schema, b) break it up into two separate migrations (first create the column, then backfill data, then make it non-nullable) or c) generate a migration and backfill the data into the column before making the column required. 

c would be done through customization of the SQL file. It's very easy to skip this step if you are not applying the migrations.

**Reply by albertoperdomo** [reply]
>I think it would be a lot easier to be able to, instead of editing the schema.prisma file manually, just update your database schema using SQL, and after that re-running prisma introspect to automatically update the schema.prisma file with only the changes that were added to the database, what is much less error prone than editing the file manually.

Also be aware that Prisma Schema Language does not cover 100% of the database features.

**Reply by carlos-moreira** [reply]
@albertoperdomo I agree with you that we can not trust 100% the migration generated by prisma to reflect exactly the changes that were made on the database, even the docs says that here: https://www.prisma.io/docs/concepts/components/prisma-migrate/prisma-migrate-flows

> You can choose to change the behavior of a migration when you create it. For example, Prisma Migrate cannot automatically detect that you want to rename a field - instead, it will add a new field and remove the old one.

But as I'm executing on the step 3 the command ```prisma migrate dev --create-only --preview-feature``` I still have time to review the generated migration file and check if the queries are reflecting exactly the changes that were made on the database.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
